models/settings_helpers.go

```diff
@@ -44,39 +46,154 @@ func GetSettings(q reform.DBTX) (*Settings, error) {
+	if params.SSHKey != "" {
+		settings.SSHKey = params.SSHKey
+	}
+	if params.AlertManagerURL != "" {
+		settings.AlertManagerURL = params.AlertManagerURL
+	}
+	if params.RemoveAlertManagerURL {
+		settings.AlertManagerURL = ""
+	}
+
+	err = SaveSettings(q, settings)
+	if err != nil {
+		return nil, err
+	}
+	return settings, nil
+}
+
+// ValidateSettings validates settings changes.
+func ValidateSettings(params *ChangeSettingsParams) error {
+	if params.EnableTelemetry && params.DisableTelemetry {
+		return errors.New("Both enable_telemetry and disable_telemetry are present.")
+	}
+
+	checkCases := []struct {
+		dur       time.Duration
+		fieldName string
 	}{
-		{dur: s.MetricsResolutions.HR, name: "hr"},
-		{dur: s.MetricsResolutions.MR, name: "mr"},
-		{dur: s.MetricsResolutions.LR, name: "lr"},
-	} {
-		if pair.dur < time.Second {
-			return status.Error(codes.InvalidArgument, pair.name+": minimal resolution is 1s")
+		{params.MetricsResolutions.HR, "hr"},
+		{params.MetricsResolutions.MR, "mr"},
+		{params.MetricsResolutions.LR, "lr"},
+	}
+	for _, v := range checkCases {
+		if v.dur == 0 {
+			continue
 		}
-		if pair.dur.Truncate(time.Second) != pair.dur {
-			return status.Error(codes.InvalidArgument, pair.name+": should be a natural number of seconds")
+
+		if _, err := validators.ValidateMetricResolution(v.dur); err != nil {
+			switch err.(type) {
+			case validators.DurationNotAllowedError:
+				return errors.New(fmt.Sprintf("%s: should be a natural number of seconds", v.fieldName))
+			case validators.MinDurationError:
+				return errors.New(fmt.Sprintf("%s: minimal resolution is 1s", v.fieldName))
+			default:
+				return errors.New(fmt.Sprintf("%s: unknown error for", v.fieldName))
+			}
 		}
 	}
 
-	if s.DataRetention < 24*time.Hour {
-		return status.Error(codes.InvalidArgument, "data_retention: minimal resolution is 24h")
-	}
-	if s.DataRetention.Truncate(24*time.Hour) != s.DataRetention {
-		return status.Error(codes.InvalidArgument, "data_retention: should be a natural number of days")
+	if params.DataRetention != 0 {
+		if _, err := validators.ValidateDataRetention(params.DataRetention); err != nil {
+			switch err.(type) {
+			case validators.DurationNotAllowedError:
+				return errors.New("data_retention: should be a natural number of days")
+			case validators.MinDurationError:
+				return errors.New("data_retention: minimal resolution is 24h")
+			default:
+				return errors.New("data_retention: unknown error")
+			}
+		}
 	}
 
 	var err error
-	if s.AWSPartitions, err = validateAWSPartitions(s.AWSPartitions); err != nil {
+	if err = validators.ValidateAWSPartitions(params.AWSPartitions); err != nil {
 		return err
 	}
 
+	if params.AlertManagerURL != "" {
+		if params.RemoveAlertManagerURL {
+			return errors.New("Both alert_manager_url and remove_alert_manager_url are present.")
+		}
+
+		// custom validation for typical error that is not handled well by url.Parse
+		if !strings.Contains(params.AlertManagerURL, "//") {
```

Comment:
`example.org//` will make this to fail

Reply:
we have check for this kind of urls below and in tests